lib/semian/sysv_shared_memory.rb

```diff
@@ -0,0 +1,50 @@
+module Semian
+  module SysVSharedMemory #:nodoc:
+    def self.included(base)
+      def base.do_with_sync(*names)
+        names.each do |name|
+          new_name = "#{name}_inner".freeze.to_sym
+          alias_method new_name, name
+          private new_name
+          define_method(name) do |*args, &block|
+            synchronize do
+              method(new_name).call(*args, &block)
+            end
+          end
+        end
+      end
+    end
```

Comment:
I'll try to put what I meant into code, I think all you need to do is make sure you don't try to grab the mutex again if you nest (with the mutex here being the shared memory one, I'm just using a pure Ruby example):

```ruby
class A
  def value
    synchronize { @value }
  end

  private
  def synchronize
    yield
  end
end

class B < A
  def initialize
    @mutex = Mutex.new
  end

  def increment(val = 1)
    synchronize { self.value += val }
  end 

  private
  def synchronize
    return yield if @nested
    @mutex.synchronize {
       @nested = true
       yield 
     }
  end
end 
```